### PR TITLE
Skip attention mask materialization in RL training

### PR DESCRIPTION
**Pull request by tdene**
- [x] I, the PR author, have personally reviewed every line of this PR.

# What does this PR do?
Adds a `create_attention_mask` argument to `get_ltor_masks_and_position_ids` that can be toggled to `False`, skipping the creation of the attention mask when it's unnecessary.

## Issue tracking

For PRs from open-source community contributors:

- **New features**: a linked issue is **required**. Please open a [feature request](https://github.com/NVIDIA/Megatron-LM/issues/new?template=feature_request.md) and reference it here before submitting the PR.
- **Small updates (bug fixes, minor improvements)**: a linked issue is **recommended** and will accelerate the PR review process.

Linked issue: 

### Pre-checks

- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment @NVIDIA/mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

### CODE REVIEW

File: megatron/training/utils/common_utils.py [hunk 1]
```diff
@@ -429,7 +435,8 @@ def get_ltor_masks_and_position_ids(data,
                     prev_index = i + 1
 
     # Convert attention mask to binary:
-    attention_mask = attention_mask < 0.5
+    if create_attention_mask:
```
Comment (yobibyte) [inline megatron/training/utils/common_utils.py:456]: Why not do this above 399?

File: megatron/training/utils/common_utils.py [hunk 2]
```diff
@@ -377,20 +377,26 @@ def get_ltor_masks_and_position_ids(data,
                                     reset_position_ids,
                                     reset_attention_mask,
                                     eod_mask_loss,
-                                    pad_mask_loss):
+                                    pad_mask_loss,
+                                    create_attention_mask=True):
     """Build masks and position id for left to right model."""
+    assert create_attention_mask or not reset_attention_mask, \
```
Comment (yobibyte) [inline megatron/training/utils/common_utils.py:401]: I think we need a docstring now explaining what reset/create attention mask do and why we need both.

File: megatron/training/utils/common_utils.py [hunk 3]
```diff
@@ -429,7 +435,8 @@ def get_ltor_masks_and_position_ids(data,
                     prev_index = i + 1
 
     # Convert attention mask to binary:
-    attention_mask = attention_mask < 0.5
+    if create_attention_mask:
+        attention_mask = attention_mask < 0.5
```
Comment (yobibyte) [inline megatron/training/utils/common_utils.py:457]: why not do this in 399?

**Review by yobibyte — COMMENTED** [review]
I would run some 20 steps of an environment before/after for this to compare, but I do not have an argument why. Just intuition. Also left some comments.

File: megatron/training/utils/common_utils.py (hunk 1 above)
Comment (tdene) [reply megatron/training/utils/common_utils.py:456]: The code between 399 and 438 mutates the `attention_mask`.

This is also the established code pattern that `_get_ltor_masks_and_position_ids` in `megatron/core/datasets/gpt_dataset.py` (around L765) decided upon. Code consistency.

File: megatron/training/utils/common_utils.py (hunk 2 above)
Comment (tdene) [reply megatron/training/utils/common_utils.py:401]: Done, what do you think?

File: megatron/training/utils/common_utils.py (hunk 3 above)
Comment (tdene) [reply megatron/training/utils/common_utils.py:457]: Duplicate, resolving.

**Comment by tdene** [comment]
> I would run some 20 steps of an environment before/after for this to compare, but I do not have an argument why. Just intuition. Also left some comments.

I did that here, although the "after" in this run was multiple changes packaged together: https://wandb.ai/adlr/megatron-rl?nw=9vg5y1477c9

Do you want an actual before/after without any confounding variables, or were you asking for just a sanity check?

**Comment by yobibyte** [comment]
> > I would run some 20 steps of an environment before/after for this to compare, but I do not have an argument why. Just intuition. Also left some comments.
> 
> I did that here, although the "after" in this run was multiple changes packaged together: https://wandb.ai/adlr/megatron-rl?nw=9vg5y1477c9
> 
> Do you want an actual before/after without any confounding variables, or were you asking for just a sanity check?

sanity check, looks ok

**Review by yobibyte — APPROVED** [review]
LGTM

File: megatron/training/utils/common_utils.py (hunk 2 above)
Comment (yobibyte) [reply megatron/training/utils/common_utils.py:401]: looks ok

**Review by deepakn94 — APPROVED** [review]

**Review by asolergi-nv — APPROVED** [review]
